### PR TITLE
SI-7291: Don't throw exceptions while encountering diverging expansion.

### PR DESCRIPTION
**Pull request by hubertp**
Since we don't throw exceptions for normal errors it was a bit odd that we don't do that for DivergingImplicit. As SI-7291 shows, the logic behind catching/throwing exception was broken for divergence. Instead of patching it, I rewrote the mechanism so that we now another SearchFailure type related to diverging expansion, similar to ambiguous implicit scenario.
The logic to prevent diverging expansion from stopping the search had to be slightly adapted but works as usual.

The upside is that we don't have to catch diverging implicit for example in the presentation compiler which was again showing that something was utterly broken with the exception approach.

Review by @retronym.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 1]
```diff
@@ -59,6 +59,25 @@ trait ContextErrors {
...
+  // becasue we don't modify implicits' infos.
+  case class DivergentImplicitTypeError(tree: Tree, pt0: Type, sym: Symbol) extends AbsTypeError {
+    def errPos: Position = tree.pos
+    def errMsg: String   = errMsgForPt(pt0)
+    def kind = ErrorKinds.Divergent
+    def withPt(pt: Type): AbsTypeError =
+      NormalTypeError(tree, errMsgForPt(pt), kind)
+    private def errMsgForPt(pt: Type) = 
+      "diverging implicit expansion for type "+pt+"\nstarting with "+
+      sym.fullLocationString
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala]: string interpolation is preferred.

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 2]
```diff
@@ -820,11 +814,19 @@ trait Implicits {
...
         case i :: is  =>
-          def tryImplicitInfo(i: ImplicitInfo) =
-            try typedImplicit(i, ptChecked = true, isLocal)
-            catch divergenceHandler
+          def tryImplicitInfo(i: ImplicitInfo) = {
+            val search = typedImplicit(i, ptChecked = true, isLocal)
+            if (search.isDivergent && divergenceCountdown > 0) {
+              divergenceCountdown -= 1
+              divergentSym = i.sym
+              log("discarding divergent implicit during implicit search")
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: How about "discarding divergent implicit $divergentSym".

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 3]
```diff
@@ -783,16 +785,8 @@ trait Implicits {
...
-      private val divergenceHandler: PartialFunction[Throwable, SearchResult] = {
-        var remaining = 1;
-        { case x: DivergentImplicit if remaining > 0 =>
-            remaining -= 1
-            divergence = true
-            log("discarding divergent implicit during implicit search")
-            SearchFailure
-        }
-      }
+      private var divergentSym: Symbol = null
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: Can you comment that you are using null as a marker?

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 4]
```diff
@@ -820,11 +814,19 @@ trait Implicits {
...
-          def tryImplicitInfo(i: ImplicitInfo) =
-            try typedImplicit(i, ptChecked = true, isLocal)
-            catch divergenceHandler
+          def tryImplicitInfo(i: ImplicitInfo) = {
+            val search = typedImplicit(i, ptChecked = true, isLocal)
+            if (search.isDivergent && divergenceCountdown > 0) {
+              divergenceCountdown -= 1
+              divergentSym = i.sym
+              log("discarding divergent implicit during implicit search")
+              SearchFailure
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: Might be nicer to encapsulate the mutable fields for countdown/recovery logic.

``` scala
object DivergentImplicitRecovery {
  private var sym: Symbol = _
  private var countdown: Int = 1
  def apply(search: SearchResult): SearchResult = ...
}

...

val search = typedImplicit(i, ptChecked = true, isLocal)
DivergentImplicitRecovery(search)
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 5]
```diff
@@ -151,14 +151,14 @@ trait Typers extends Adaptations with Tags {
             mkArg = gen.mkNamedArg // don't pass the default argument (if any) here, but start emitting named arguments for the following args
             if (!param.hasDefault && !paramFailed) {
               context.errBuffer.find(_.kind == ErrorKinds.Divergent) match {
-                case Some(divergentImplicit) =>
+                case Some(divergent: DivergentImplicitTypeError) =>
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Seems a bit redundant to have both `ErrorKind.Divergent` and the type `DivergentImplicitTypeError`. Maybe we can do with one of them, and have a `collect` method for errors.

**Comment by paulp** [comment]
Here is a more efficient version of normalizeFirstParent. It avoids duplicating the isTraitRef test, and doesn't create new cons cells unless they are needed. (When ys == x :: xs and you return x :: xs instead of ys, you allocate unnecessarily.) This is untested, even to the extent of compiling it, but I'm unreasonably optimistic it will work.

```
    private def normalizeFirstParent(parents: List[Tree]): List[Tree] = {
      @annotation.tailrec def explode(parents: List[Tree]): List[Tree] = {
        val supertpt :: rest = parents // parents is always non-empty here - it only grows
        if (treeInfo isTraitRef supertpt) {
          val supertpt1  = typedType(supertpt)
          def supersuper = TypeTree(supertpt1.tpe.firstParent) setPos supertpt.pos.focus
          if (supertpt1.isErrorTyped) rest
          else explode(supersuper :: supertpt1 :: rest)
        }
        else {
          if (supertpt.tpe.typeSymbol == AnyClass)
            supertpt setType AnyRefTpe

          parents
        }
      }
      if (parents.isEmpty) Nil else explode(parents)
    }
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 5 above)
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yeah, I did think about it but then a) we would have different behaviour between lazy SearchFailures b) we still do have to provide more specialized parameter type (see a couple of lines below) anyway for better error messages and unfortunately one cannot do it earlier c) here we handle the "no found implicit" and "diverging expansion..." but there are situations where we don't do it and simply rely on the fact that buffering the errors will do the right job.
I can definitely agree that the situation is not ideal but then choosing between any of the two methods only is also inconvenient.  Let me see if I can simplify it. I am not sure I get your `collect` proposition, could you elaborate?

**Comment by hubertp** [comment]
Sorry to kill your optimism but this version won't work with value classes (that's the reason why there was a duplicate isTraitRef test). Allocation hint is still useful, though. Thanks.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 5 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]:
``` scala
val x: Option[DTE] = context.errBuffer.collectFirst { 
  case dte: DivergentImplicitTypeError => dte
}
```

Basically, should we get rid of `ErrKind` and just use the type of the error? But we could (should?) do this on a followup change, to help me get my `Context` refactoring in.

Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: It's tricky. I tried to change it by basically relying only on (Diverging)SearchResult but the issue becomes with having precise error message - the divergent symbol is lost by the time we want to finally report the error message. The only way to recover it is through the context buffer.

**Comment by paulp** [comment]
@hubertp Can you explain? Why doesn't it work for value classes? I am confident whatever the reason is, I can modify this code to cover it without duplicating the test.

**Comment by paulp** [comment]
Also, please limit commit message line lengths to 72 characters, as noted (with inadequate emphasis) in the pull request policy link to http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html . 364 slightly exceeds the limit.

**Comment by hubertp** [comment]
@paulp I should have been more precise. There are bootstrapping issues with AnyVal in that version. I didn't mean it wasn't possible at all, just that there were problems with the version you proposed. Let me tinker it a little bit more.

Updated wrt to comments, rebased against current master that has the new context impl.

**Comment by paulp** [comment]
@hubertp I tried it out and did encounter the discussed bootstrapping problems. I'm a little confused about why, but the fact that there's a genuine obstacle is enough for me. This LGTM then.

**Comment by paulp** [comment]
I'm going to merge this because it sounds like it will save #2406.

**Comment by xeno-by** [comment]
May I port this to 2.10.x?

**Comment by paulp** [comment]
No problem that I can see.

**Comment by hubertp** [comment]
Were Jason's context changes backported to 2.10? That can be problematic for you, if not.

**Comment by paulp** [comment]
@hubertp I'm curious what is the story here: https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos-per-commit/1949/consoleText

It fails the same way as before ("ERROR: Right(scala.tools.nsc.typechecker.DivergentImplicit$)") and it includes this PR. I got the impression one or more someones thought it wouldn't happen like that.

**Comment by hubertp** [comment]
@paulp I guess I am seeing ghosts. DivergentImplicit is gone, seriously. Let me have a closer look.

**Pull request merged**